### ISSUE TITLE
docs: cleaning up after failed topology operation

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#13122]
After a failed topology operation, like bootstrap / decommission / removenode, the cluster might contain a garbage entry in either token ring or group 0. This entry can be cleaned-up by executing `removenode` on any other node, pointing to the node that failed to bootstrap or leave the cluster.

We need to document this.

**Comment by kbr-scylla** [comment]
Needs to be included in 5.2 documentation.

**Comment by kbr-scylla** [comment]
It might be impossible to perform other topology operations until the garbage entry is removed. Observed #13120, where a new node couldn't bootstrap because of an earlier failed decommission (the node which tried to decommission managed to leave the token ring, but not group 0) - it tried to contact the node which failed to decommission.

Not sure if it's good or bad. It might be good that we don't allow further topology operations until everything is clean. But we should still understand why exactly this happens. For example, will replace operation also be impossible, if we try to replace a node which failed a decommission earlier? Should it be impossible?

**Comment by vladzcloudius** [comment]
@kbr-scylla Could you, please, explain the motivation a little more to me?
AFAIR we didn't persist the joining node info on disk and gossiper had a logic to clean its internal (in-memory) state after failures in question before Raft.

With Raft I was expecting the situation to be even better since with Raft by design a topology state change was supposed to be handle as a single transaction including errors handling and the following cleanup.

However based on this issue it sounds like we fail to achieve so, which looks like a Raft or new topology change implementation bug to me - and not a Doc issue.

It even looks like a regression compared to a pre-Raft gossiper based implementation.

Please, comment.

**Comment by kbr-scylla** [comment]
>  AFAIR we didn't persist the joining node info on disk and gossiper had a logic to clean its internal (in-memory) state after failures in question before Raft.

For joining perhaps it's true that without Raft Scylla will always clean-up automatically without Raft.

For decommission it may happen with a well-timed crash that we end up in a state where the node is broken and cannot restart - even with Raft disabled.

> With Raft I was expecting the situation to be even better since with Raft by design a topology state change was supposed to be handle as a single transaction including errors handling and the following cleanup.

We don't have Raft-based topology changes yet. Only Raft-based schema changes.

For topology changes currently Raft makes the situation worse indeed. That's because during topology changes, additionally to maintaining the token ring information, we also maintain group 0 membership. For example on bootstrap we may crash after joining group 0 but before finishing token ring transition, and then we'll end up with this partially joined member.

Yes it is a regression.

cc @kostja

**Comment by vladzcloudius** [comment]
> For decommission it may happen with a well-timed crash that we end up in a state where the node is broken and cannot restart - even with Raft disabled.

Could you, please, add more specific details about the reproducer of what you are describing?
cc @asias

**Comment by kbr-scylla** [comment]
> @kbr-scylla Could you, please, add more specific details about the reproducer of what you are describing without Raft in the picture?

I inserted this `abort` on the decommission path:
```diff
diff --git a/service/storage_service.cc b/service/storage_service.cc
index de78c7b141..e91149446d 100644
--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -3150,6 +3150,7 @@ future<> storage_service::send_replication_notification(inet_address remote) {
 
 future<> storage_service::leave_ring() {
     co_await _sys_ks.local().set_bootstrap_state(db::system_keyspace::bootstrap_state::NEEDS_BOOTSTRAP);
+    std::abort();
     co_await mutate_token_metadata([this] (mutable_token_metadata_ptr tmptr) {
         auto endpoint = get_broadcast_address();
         tmptr->remove_endpoint(endpoint);
```

The node will refuse to restart after this - it will try to bootstrap and notice that there is a node with the same IP. (I didn't check what happens when it changes the IP between restarts, it could be even worse.)

Credit to @aleksbykov who reproduced it by checking what happens when you crash during decommission at different points.

**Comment by vladzcloudius** [comment]
> > @kbr-scylla Could you, please, add more specific details about the reproducer of what you are describing without Raft in the picture?
> 
> I inserted this `abort` on the decommission path:
> 
> ```diff
> diff --git a/service/storage_service.cc b/service/storage_service.cc
> index de78c7b141..e91149446d 100644
> --- a/service/storage_service.cc
> +++ b/service/storage_service.cc
> @@ -3150,6 +3150,7 @@ future<> storage_service::send_replication_notification(inet_address remote) {
>  
>  future<> storage_service::leave_ring() {
>      co_await _sys_ks.local().set_bootstrap_state(db::system_keyspace::bootstrap_state::NEEDS_BOOTSTRAP);
> +    std::abort();
>      co_await mutate_token_metadata([this] (mutable_token_metadata_ptr tmptr) {
>          auto endpoint = get_broadcast_address();
>          tmptr->remove_endpoint(endpoint);
> ```
> 
> The node will refuse to restart after this - it will try to bootstrap and notice that there is a node with the same IP. (I didn't check what happens when it changes the IP between restarts, it could be even worse.)
> 
> Credit to @aleksbykov who reproduced it by checking what happens when you crash during decommission at different points.

@kbr-scylla IMO this sounds like a bug - not a documentation issue.
@asias thoughts?

**Comment by kostja** [comment]
@kbr-scylla Apparently  we should auto-expire a node from group0 which is not part of the topology - at least until gleb's topology changes on raft are in.

**Comment by kbr-scylla** [comment]
> @kbr-scylla Apparently we should auto-expire a node from group0 which is not part of the topology - at least until gleb's topology changes on raft are in.

The time for such decisions for 5.2 has passed.
We were aware of the problem and consciously decided not to deal with it.
I proposed that we at least make it possible to remove such garbage nodes manually: https://docs.google.com/document/d/1Gvy2_ZezCa2XY5li05umOznIwjvIYFGmYpeZQg5RbZM/edit
the proposal was reviewed by multiple people, and I implemented it https://github.com/scylladb/scylladb/pull/12502 and it got merged.

The best thing we can do for 5.2 is to document the procedure. I sent a PR there: https://github.com/scylladb/scylladb/pull/13186

**Comment by kostja** [comment]
Let's separate the current state, which, to a large extent, was present for a long time (although I agree that in some cases Raft makes it worse, which cases - we will document explicitly), and improving the situation.

The ultimate improvement is raft based topology changes. For 5.3, however, we could extend nodetool to show this condition more explicitly (e.g. in something like extendedstatus) and prune zombie nodes automatically (nodetool prunedead). Let's file a ticket and triage it? 
Meanwhile I think Kamil's pull request is adding valuable information to the docs and should make its way in.

**Comment by vladzcloudius** [comment]
> Let's separate the current state, which, to a large extent, was present for a long time (although I agree that in some cases Raft makes it worse, which cases - we will document explicitly), and improving the situation.
> 
> The ultimate improvement is raft based topology changes. For 5.3, however, we could extend nodetool to show this condition more explicitly (e.g. in something like extendedstatus) and prune zombie nodes automatically (nodetool prunedead). Let's file a ticket and triage it? Meanwhile I think Kamil's pull request is adding valuable information to the docs and should make its way in.

Let's agree to disagree, @kostja.
IMO "zombi nodes" are supposed to be cleaned up by scylla automatically. Users don't need to know about their existence. Nor any additional `nodetool` commands are supposed to be needed in their context.

And AFAIK, today it *is* (was?) the case. 

The way I see it (and I'm still wondering what @asias would tell on that) you, guys, introduced a regression and regressions are one of the worst case of bugs.

And as such the corresponding GH issue is good-enough "documentation" IMO.

Are we going to turn every GH issue about a bug into a Documentation page? I hope not.

**Comment by kostja** [comment]
> And AFAIK, today it _is_ (was?) the case.
> 
> The way I see it (and I'm still wondering what @asias would tell on that) you, guys, introduced a regression and regressions are one of the worst case of bugs.
> 
> And as such the corresponding GH issue is good-enough "documentation" IMO.
> 
> Are we going to turn every GH issue about a bug into a Documentation page? I hope not.

It is more complex than that.

Previously, adding and removing nodes did not require a majority. So it was possible to proceed with a new topology operation even if the previous failed, and the cluster was in a degraded state. The dead node would indeed be cleared after a 20min timeout, but that would happen already after a new node was added. If the node wasn't really dead, but declared dead by gossiper by mistake, that would easily lead to consistency level violations.
Even if we start automatically removing a node from raft group 0 (I suggested it could be an interim improvement - but incomplete - see below) - it would still not address this "regression", as you call it, since under no circumstances we could allow forward progress within a minority.

Now, let's consider how big is this regression. In Raft, nodes are added as non-voters, and converted to voters only after the streaming is over. So join is not introducing a regression. A decommission, killed at the right spot in the process, does. On  the other hand, you could easily find a spot in the previous procedure where you would get the same outcome - Kamil pointed out a case.  

Third, there are numerous places in the code like this:
 ```                                                                                                   
            if (tmptr->is_normal_token_owner(get_broadcast_address())) {                            
                throw std::runtime_error("This node is already a member of the token ring; bootstrap aborted. (If replacing a dead node, remove the old one from the ring first.)");
```

In other words, your statement that previously a cleanup would happen automatically and we would be able to make forward progress is incorrect, or at least not quite giving the complete picture.

Needless to say this code has nearly no test coverage, and no documentation - we do not ever officially guarantee that a failed decommission is cleaned up after automatically. 

Software quality starts with the docs, we must be truthful about the guarantees we provide. A bug is first of all is inconsistency between the documented and actual behavior of the program. 

We have already acknowledged the issue and this is why we're building strongly consistent topology changes. A centralized topology change coordinator will make *reliable* automatic failure handling possible. But even then, Gleb's series for it still do not implement it - so I can't promise it will land in 5.3, even in experimental mode.

**Comment by asias** [comment]
> > @kbr-scylla Could you, please, add more specific details about the reproducer of what you are describing without Raft in the picture?
> 
> I inserted this `abort` on the decommission path:
> 
> ```diff
> diff --git a/service/storage_service.cc b/service/storage_service.cc
> index de78c7b141..e91149446d 100644
> --- a/service/storage_service.cc
> +++ b/service/storage_service.cc
> @@ -3150,6 +3150,7 @@ future<> storage_service::send_replication_notification(inet_address remote) {
>  
>  future<> storage_service::leave_ring() {
>      co_await _sys_ks.local().set_bootstrap_state(db::system_keyspace::bootstrap_state::NEEDS_BOOTSTRAP);
> +    std::abort();
>      co_await mutate_token_metadata([this] (mutable_token_metadata_ptr tmptr) {
>          auto endpoint = get_broadcast_address();
>          tmptr->remove_endpoint(endpoint);
> ```
> 
> The node will refuse to restart after this - it will try to bootstrap and notice that there is a node with the same IP. (I didn't check what happens when it changes the IP between restarts, it could be even worse.)


This is actually a good thing that restart with the same ip would fail. It is bad that we could not restart the node hasn't finished the decommission completely in this rare case.  But it is better than we could restart the nodes successfully, but some other nodes have seen the LEFT status and think the node is removed.  For example, we move the code like:

```
advertise LEFT
std::abort()
set NEEDS_BOOTSTRAP
```

This decommission node will restart sucessfully and think it is still in the ring while the other nodes think it has left. 

If the node has changed ip address, it will join the node as if it is a new node using different tokens. The old node will stay as normal but down. We can only run removenode to remove it from the cluster. Not that bad. 


> 
> Credit to @aleksbykov who reproduced it by checking what happens when you crash during decommission at different points.

**Comment by avikivity** [comment]
Already backported to all vulnerable branches, removing "Backport candidate" label.
